src/Microsoft.AspNet.Mvc.Razor/Compilation/ExpressionRewriter.cs

```diff
@@ -0,0 +1,161 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq.Expressions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.AspNet.Mvc.Razor.Compilation
+{
+    /// <summary>
+    /// An expression rewriter which can hoist a simple expression lambda into a private field.
+    /// </summary>
+    public class ExpressionRewriter : CSharpSyntaxRewriter
+    {
+        private static readonly string FieldNameTemplate = "__h{0}";
+
+        public ExpressionRewriter(SemanticModel semanticModel)
+        {
+            SemanticModel = semanticModel;
+
+            Expressions = new List<KeyValuePair<SimpleLambdaExpressionSyntax, IdentifierNameSyntax>>();
+        }
+
+        private SemanticModel SemanticModel { get; }
+
+        private List<KeyValuePair<SimpleLambdaExpressionSyntax, IdentifierNameSyntax>> Expressions { get; }
+
+        public override SyntaxNode VisitClassDeclaration(ClassDeclarationSyntax node)
+        {
+            Expressions.Clear();
```

Comment:
If this is a concern we can make it more robust, or at least we could add some validation that it's not being used that way. I left it simple on purpose since there's no way to to run this on non-Razor code.